exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/list/FlatListProof.java

```diff
@@ -0,0 +1,440 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.common.proofs.list;
+
+import static com.exonum.binding.common.hash.Funnels.hashCodeFunnel;
+import static com.exonum.binding.common.hash.Hashing.sha256;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.math.BigIntegerMath.log2;
+import static java.util.stream.Collectors.toMap;
+
+import com.exonum.binding.common.hash.HashCode;
+import com.exonum.binding.common.hash.Hasher;
+import com.exonum.binding.common.hash.Hashing;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Functions;
+import java.math.BigInteger;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.BinaryOperator;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * A flat list proof. It proves that certain elements are present in a proof list
+ * of a certain size.
+ */
+class FlatListProof {
+  /*
+    Proof lists are represented as BSTs, where all leaf elements are at the same height.
+    Here is a tree for a three-element proof list:
+
+     H — height of nodes at each level
+     2        o
+            /   \
+     1    o       o
+         / \     /
+     0  e   e   e
+
+    Proofs for elements of proof lists include the list size; the requested elements;
+    and hashes of all sub-trees that are adjacent to the paths from the root node to the leaf nodes
+    containing the elements.
+
+    Here is a proof for the element at index 1 from the list shown above:
+
+     H
+     2        o
+            /   \
+     1    o       h
+         / \
+     0  h   e
+     ________________
+     Legend:
+       e — element proof entry (ListProofElementEntry)
+       h — hashed proof entry (ListProofHashedEntry)
+       o — 'virtual' node — not present in the proof, inferred during verification.
+           Shown mostly to communicate the tree structure of the proof.
+
+    See also: https://wiki.bf.local/display/EXN/Flat+list+proofs
```

Comment:
It will be public soon, but this comment is not a Javadoc, it is an implementation comment, and having the docs available and discoverable to _some_ developers (actually — most 🙃) is better than to none.

I could mark it with `[internal]` if needed

Reply:
It's good as it is, thanks - indeed it's not a Javadoc.